test/ioctl_zeroing.cpp

```diff
@@ -0,0 +1,107 @@
+// Some ioctls have an output_size_bytes input value. When the actual output
+// data is smaller than output_size_bytes, the remainder must be zero-filled.
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <cerrno>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+
+#include <sys/ioctl.h>
+
+#include "aligned_allocator.h"
+#include "devfd.h"
+#include "enumeration.h"
+#include "ioctl.h"
+#include "test_failure.h"
+#include "util.h"
+
+namespace
+{
+
+#define CHECK_IOCTL_ZEROING(fd, ioctl_name, ioctl_data) CheckIoctlZeroing(fd, ioctl_name, #ioctl_name, ioctl_data)
+
+template <class IoctlData>
+void CheckIoctlZeroing(int fd, unsigned long ioctl_code, const char *ioctl_name, const IoctlData &ioctl_data)
+{
+    std::vector<unsigned char, AlignedAllocator<unsigned char, alignof(IoctlData)>> buf;
+    buf.resize(sizeof(IoctlData::in) + page_size(), 0xFF);
```

Comment:
Why `sizeof(IoctlData::in)` and not `sizeof(IoctlData)`?

Reply:
Because in.output_size_bytes is how many bytes are available for output starting at &arg->out, i.e sizeof(IoctlData::in) bytes into the structure.

But there could be padding between in & out, so actually it must be offsetof(IoctlData, out).